src/reflect/scala/reflect/internal/Scopes.scala

```diff
@@ -282,6 +282,30 @@ trait Scopes extends api.Scopes { self: SymbolTable =>
       }
     }
 
+    final def lookupSymbolEntry(sym: Symbol): ScopeEntry = {
+      var e = lookupEntry(sym.name)
+      while (e ne null) {
+        if (e.sym == sym) return e
+        e = lookupNextEntry(e)
+      }
+      null
+    }
+
+    final def lookupCompanion(original: Symbol): Symbol = {
+      lookupSymbolEntry(original) match {
+        case null =>
+        case entry if entry.sym == original =>
+          entry.owner.lookupAllEntries(original.name.companionName)
+            .filter(_.owner eq entry.owner)
+            .filter(x => original.isTerm || x.sym.hasModuleFlag).toList match {
```

Comment:
I've re-added the `isCodefinedWith` check, moved some code around, and added a comment for the `hasModuleFlag` check.